clang-tools-extra/clang-tidy/bugprone/CastingThroughVoidCheck.cpp

```diff
@@ -0,0 +1,50 @@
+//===--- CastingThroughVoidCheck.cpp - clang-tidy -------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "CastingThroughVoidCheck.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/Type.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "llvm/ADT/StringSet.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+void CastingThroughVoidCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      explicitCastExpr(
+          hasDestinationType(qualType(pointsTo(qualType(unless(voidType()))))
+                                 .bind("target_type")),
+          hasSourceExpression(
+              explicitCastExpr(hasSourceExpression(expr(
+                                   hasType(qualType().bind("source_type")))),
+                               hasDestinationType(pointsTo(voidType())))
+                  .bind("cast")),
+          unless(builtinBitCastExpr())),
```

Comment:
any reason why we need to exclude this ? if we got some warnings from std, then better would be simply to exclude entire std with isInStdNamespace or just parent function or just write some simple matcher here BuildinBitCastExpr::classOf or inver this and put anyOf(cxxStaticCastExpr(), cxxConstCastExpr(), ...)

Reply:
I'd rather like to avoid excluding the `std` namespace wholesale, since clang-tidy is used by both libc++ and the MSVC STL.

Reply:
I thought about it again and maybe we shouldn't exclude this case.